### PR TITLE
Grab bag of UTF8 string support in IDE features

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes most of https://github.com/dotnet/roslyn/issues/60582
Relates to https://github.com/dotnet/roslyn/issues/58848

Also fixed a couple of raw string things that were missed. If the review is too annoying, each commit is a separate IDE feature.

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs [hunk 1]
```diff
@@ -417,7 +417,12 @@ private static bool AtEndOfIncompleteStringOrCharLiteral(SyntaxToken token, int
                 token = token.GetPreviousToken(includeSkipped: true, includeDirectives: true);
             }
 
-            if (token.Kind() is SyntaxKind.StringLiteralToken or SyntaxKind.SingleLineRawStringLiteralToken or SyntaxKind.MultiLineRawStringLiteralToken)
+            if (token.IsKind(SyntaxKind.StringLiteralToken,
+                             SyntaxKind.SingleLineRawStringLiteralToken,
+                             SyntaxKind.MultiLineRawStringLiteralToken,
```
Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs:422]: I see these already existed before your change. But I'm confused with it when reading this comment:

https://github.com/dotnet/roslyn/blob/c8ebc8682889b395fcb84c85bf4ff54577377d26/src/Analyzers/CSharp/CodeFixes/ConvertNamespace/ConvertNamespaceTransform.cs#L126-L129

File: src/VisualStudio/CSharp/Impl/LanguageService/CSharpHelpContextService.cs [hunk 2]
```diff
@@ -133,6 +133,21 @@ private static bool TryGetTextForSpecialCharacters(SyntaxToken token, [NotNullWh
...
+                text = Keyword("UTF8StringLiteral");
+                return true;
+            }
+
+            if (token.IsKind(SyntaxKind.SingleLineRawStringLiteralToken) ||
+                token.IsKind(SyntaxKind.MultiLineRawStringLiteralToken))
+            {
+                text = Keyword("RawStringLiteral");
+                return true;
+            }
```
Comment (Youssef1313) [inline src/VisualStudio/CSharp/Impl/LanguageService/CSharpHelpContextService.cs:149]: @BillWagner FYI to use those f1 keywords when documenting the features

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs (hunk 1 above)
Comment (davidwengier) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs:422]: Yes, interesting disconnect there. The comment isn't wrong, in that raw strings can be _potentially_ dedented, but definitely outside the scope of this PR.

File: src/VisualStudio/CSharp/Impl/LanguageService/CSharpHelpContextService.cs (hunk 2 above)
Comment (davidwengier) [reply src/VisualStudio/CSharp/Impl/LanguageService/CSharpHelpContextService.cs:149]: Currently F1ing on a UTF8 string goes to a "No help content found" page, so its not a big deal IMO if these keywords are specified before docs pages exist/are updated.

**Comment by JoeRobich** [comment]
Retargeting to main so we can remove main-vs-deps branch

File: src/EditorFeatures/CSharp/BraceMatching/StringLiteralBraceMatcher.cs [hunk 3]
```diff
@@ -34,18 +34,11 @@ public StringLiteralBraceMatcher()
...
-                    {
-                        return new BraceMatchingResult(
-                            new TextSpan(token.SpanStart, 1),
-                            new TextSpan(token.Span.End - 1, 1));
-                    }
+                    return GetSimpleStringBraceMatchingResult(token, 1);
+                }
+                else if (token.IsKind(SyntaxKind.UTF8StringLiteralToken))
+                {
+                    return GetSimpleStringBraceMatchingResult(token, 3);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/BraceMatching/StringLiteralBraceMatcher.cs:41]: named parameter for '1' and '3' plz.

File: src/Features/CSharp/Portable/SplitStringLiteral/SimpleStringSplitter.cs [hunk 4]
```diff
@@ -28,8 +28,14 @@ private sealed class SimpleStringSplitter : StringSplitter
...
 
             // Don't split @"" strings.  They already support directly embedding newlines.
+            // Don't split UTF8 strings if the cursor is after the quote.
             protected override bool CheckToken()
-                => !_token.IsVerbatimStringLiteral();
+                => !_token.IsVerbatimStringLiteral() && !CursorIsAfterQuotesInUTF8String();
+
+            private bool CursorIsAfterQuotesInUTF8String()
+            {
+                return _token.IsKind(SyntaxKind.UTF8StringLiteralToken) && CursorPosition >= _token.Span.End - 2;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/SplitStringLiteral/SimpleStringSplitter.cs:37]: consider things like `"u8".length`

**Comment by davidwengier** [comment]
ping @dotnet/roslyn-ide for review

File: src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs [hunk 5]
```diff
@@ -76,19 +87,25 @@ private static int GetStartOfRawStringLiteralEndDelimiter(SyntaxToken token)
         }
 
         private static bool IsAtClosingQuote(SyntaxToken token, int position)
-            => position == token.Span.End - 1 && token.Text[^1] == '"';
+            => token.Kind() switch
+            {
+                SyntaxKind.StringLiteralToken => position == token.Span.End - 1 && token.Text[^1] == '"',
+                SyntaxKind.UTF8StringLiteralToken => position == token.Span.End - 3 && token.Text is [.., '"', 'u' or 'U', '8'],
```
Comment (ryzngard) [inline src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs:93]: fancy

File: src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs [hunk 6]
```diff
@@ -76,19 +87,25 @@ private static int GetStartOfRawStringLiteralEndDelimiter(SyntaxToken token)
...
+                _ => throw ExceptionUtilities.Unreachable
+            };
 
         protected override TextExtent GetExtentOfWordFromToken(SyntaxToken token, SnapshotPoint position)
         {
-            if (token.Kind() == SyntaxKind.StringLiteralToken && IsAtClosingQuote(token, position.Position))
+            if (token.Kind() is SyntaxKind.StringLiteralToken or SyntaxKind.UTF8StringLiteralToken && IsAtClosingQuote(token, position.Position))
```
Comment (ryzngard) [inline src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs:99]: There's different uses of `token.Kind() is A or B` vs `token.IsKind(A) || token.IsKind(B)`. Can we pick one? :)

File: src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs [hunk 7]
```diff
@@ -265,6 +278,58 @@ void M()
...
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.SplitStringLiteral)]
+        public void TestMissingAfterUTF8String_2()
+        {
+            TestNotHandled(
+@"class C
+{
+    void M()
+    {
+        var v = """"u8[||];
```
Comment (ryzngard) [inline src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs:302]: Not that it should matter, but a test for `""u[||]8` would be nice

**Review by ryzngard — APPROVED** [review]
LGTM. Can you add tests to MoveToFile as well to make sure they get handled correctly? (I didn't see any in source prior to this change)

File: src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs (hunk 5 above)
Comment (davidwengier) [reply src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs:93]: If we can't use new language features, who can?

File: src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs (hunk 6 above)
Comment (davidwengier) [reply src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs:99]: I'm 99% sure I once used an overload of `IsKind` that took multiple parameters too, but then sometimes I can't find it ¯\\\_(ツ)_/¯

Comment (Youssef1313) [reply src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs:99]: https://github.com/dotnet/roslyn/blob/b0cd71f82fbada617a7cb5354f52bbc478c1ff1d/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTokenExtensions.cs#L41-L45

**Comment by davidwengier** [comment]
> Can you add tests to MoveToFile as well to make sure they get handled correctly?

Will do, but I'm curious, did you have something specific in mind for this? It would surprise me if MoveToFile was sensitive to new syntax nodes that aren't new member declarations.

**Comment by davidwengier** [comment]
Actually.. I'm struggling to work out where a test for MoveToFile would make sense. Just to be clear do you mean `MoveTypeTests`? There don't seem to be tests for other string literals or even any other specific syntax nodes, so I'm not sure what you're looking to validate.

File: src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs (hunk 6 above)
Comment (davidwengier) [reply src/EditorFeatures/CSharp/TextStructureNavigation/CSharpTextStructureNavigatorProvider.cs]: But only one has style!
SwagPoseGIF

**Pull request merged**